### ISSUE TITLE
Can't filter by traits on `ITheoryDataRow`

### ISSUE THREAD

**Issue by justindbaur** [xunit/visualstudio.xunit#426]
I had reported xunit/xunit#3030 before, on later versions I no longer get a `Catastrophic failure` but a trait filter with traits that come from `ITheoryDataRow` still don't appear to work. 

Minimal repo: https://github.com/justindbaur/xunit-v3-problem

I expect `dotnet test --filter "MyTrait=One"` to run just the theory data row with `1, 2`. But instead it finds no tests with the specified filter. 

```C#
[
    new TheoryDataRow(1, 2)
        .WithTrait("MyTrait", "One"),
    new TheoryDataRow(2, 1)
        .WithTrait("MyTrait", "Two"),
]
```

I'm using: 
```
xunit.v3: 0.6.0-pre.7
xunit.runner.visualstudio: 3.0.0-pre.49
```

**Comment by bradwilson** [comment]
I am working to reproduce the problem, but I'm running into an issue with the newest .NET SDK: https://github.com/microsoft/vstest/issues/10605

The TL;DR: it appears that `dotnet test` runs projects differently when they support Microsoft.Testing.Platform (which xUnit.net v3 projects do). This used to be behavior that you had to opt into, but now is behavior that is always enabled. Long story short, the command line options that you see in `dotnet test -?` are no longer valid, so I can't reproduce the `--filter` bug (though I suspect I know why it exists).

On the plus side, the new way to do this would be: `dotnet test -- --filter-trait "MyTrait=One"` which I have verified is working as expected. You can see in this example that `--filter` is ignored entirely now, as I run it with the old command line switch and the new command line switch:

![image](https://github.com/user-attachments/assets/cb23ecdc-4daa-484a-951a-53e06d29fc5c)

**Comment by bradwilson** [comment]
Okay, I am able to reproduce this on `8.0.403` and the good news is, there is a workaround. If you enable theory pre-enumeration, then you can filter by the trait:

![image](https://github.com/user-attachments/assets/30fd182c-13c3-44c6-8fe5-8a313eed0133)

This verifies my hypothesis. For quite a while, we have tried to optimize performance by disabling theory pre-enumeration if you were running tests from the command line, on the idea that it was unnecessary:

https://github.com/xunit/visualstudio.xunit/blob/9154f3ffa002ce582633f9e1997f2d3ad9620966/src/xunit.runner.visualstudio/VsTestRunner.cs#L210-L212

What's changed between v2 and v3 is that traits are now available via ITheoryDataRow but we never updated the code here and so when we don't pre-enumerate your theory data, we never see the traits and you can't filter by them.

The good news is, the fix is quite simple, so that's my next stop. 😄

**Comment by bradwilson** [comment]
Fix available in `3.0.0-pre.53` https://xunit.net/docs/using-ci-builds

**Comment by bradwilson** [comment]
Not quite fixed yet. 😬

**Comment by bradwilson** [comment]
Okay, but for real this time. 😂

Fix available in `3.0.0-pre.54` https://xunit.net/docs/using-ci-builds

**Comment by bradwilson** [comment]
And of course I see now my commit message says "off" instead of "on", but whatever. 💀
